File: backend/src/main/java/org/sport/backend/service/MatchService.java
package org.sport.backend.service;

import jakarta.transaction.Transactional;
import org.sport.backend.dto.base.PageResponse;
import org.sport.backend.constant.MatchStatus;
import org.sport.backend.constant.MatchType;
import org.sport.backend.dto.request.chat.DivideTeamRequest;
import org.sport.backend.dto.request.match.AutoMatchRequest;
import org.sport.backend.dto.request.match.MatchRequest;
import org.sport.backend.dto.response.match.MatchResponse;
import org.sport.backend.dto.response.payment.CheckoutResponse;

import java.time.LocalDateTime;
import java.util.UUID;

public interface MatchService {

    MatchResponse createMatch(MatchRequest request);

    @Transactional
    CheckoutResponse joinMatch(UUID matchId, Integer playerCount);

    @Transactional
    void joinByRoomCode(String roomCode, Integer playerCount);

    @Transactional
    MatchResponse autoMatch(AutoMatchRequest request);

    @Transactional
    void divideTeams(UUID matchId, DivideTeamRequest request);

    PageResponse<MatchResponse> getOpenMatches(
            int page,
            int size,
            String category,
            String keyword,
            LocalDateTime startDate,
            LocalDateTime endDate,
            MatchType matchType,
            String ward, String city);

    MatchResponse getMatchDetail(UUID matchId);

    PageResponse<MatchResponse> getAllMatches(
            int page, int size, MatchStatus status, String category, String keyword,
            LocalDateTime start, LocalDateTime end, MatchType matchType);

    PageResponse<MatchResponse> getOwnerMatchesPaged(
            int page, int size, MatchStatus status, String category,
            String keyword, LocalDateTime startDate, LocalDateTime endDate);

    PageResponse<MatchResponse> getMyMatches(int page, int size);

    PageResponse<MatchResponse> getUserMatchHistory(UUID userId, int page, int size);

    @Transactional
    void generateNextMatches();

    @Transactional
    void leaveMatch(UUID matchId);
}

File: frontend/src/service/matchService.ts
import api from '../config/axios.ts';
import type { ApiResponse } from "../types/ApiResponse.ts";
import type { MatchResponse, MatchRequest } from '../types/match';
import type {PageResponse} from "../types/post.ts";

const API_BASE_URL = '/matches';

export const matchService = {
    createMatch: async (data: MatchRequest): Promise<ApiResponse<MatchResponse>> => {
        const response = await api.post(`${API_BASE_URL}/create`, data);
        return response.data;
    },

    joinMatch: async (matchId: string): Promise<ApiResponse<void>> => {
        const response = await api.post(`${API_BASE_URL}/${matchId}/join`);
        return response.data;
    },

    getOpenMatches: async (): Promise<ApiResponse<MatchResponse[]>> => {
        const response = await api.get(`${API_BASE_URL}/open`);
        return response.data;
    },

    getMatchDetail: async (matchId: string): Promise<ApiResponse<MatchResponse>> => {
        const response = await api.get(`${API_BASE_URL}/${matchId}`);
        return response.data;
    },

    filterMatches: (
        page: number,
        size: number,
        category?: string,
        level?: string,
        keyword?: string
    ) => {
        const params: any = { page, size };
        if (category) params.category = category;
        if (level) params.level = level;
        if (keyword) params.keyword = keyword;

        return api.get(`${API_BASE_URL}/filter`, { params });
    },

    getAllMatches: async (
        page: number,
        size: number,
        status?: string,
        category?: string,
        keyword?: string,
        startDate?: string,
        endDate?: string
    ): Promise<ApiResponse<PageResponse<MatchResponse>>> => {
        const params: any = { page, size };

        if (status) params.status = status;
        if (category) params.category = category;
        if (keyword) params.keyword = keyword;
        if (startDate) params.startDate = startDate;
        if (endDate) params.endDate = endDate;

        const response = await api.get(API_BASE_URL, { params });
        return response.data;
    },

    getOwnerMatches: async (page: number, size: number): Promise<ApiResponse<PageResponse<MatchResponse>>> => {
        const response = await api.get(`${API_BASE_URL}/owner`, {
            params: { page, size }
        });
        return response.data;
    },
};

export default matchService;